consensus/polybft/consensus_runtime.go

```diff
@@ -0,0 +1,851 @@
+		return err
+	}
+
+	return nil
+}
+
+// deliverMessage receives the message vote from transport and inserts it in state db for given epoch.
+// It returns indicator whether message is processed successfully and error object if any.
+func (c *consensusRuntime) deliverMessage(msg *TransportMessage) (bool, error) {
+	epoch := c.getEpoch()
+	if epoch == nil || msg.EpochNumber < epoch.Number {
+		// Epoch metadata is undefined
+		// or received message for some of the older epochs.
+		return false, nil
+	}
+
+	if !c.isActiveValidator() {
+		return false, fmt.Errorf("validator is not among the active validator set")
+	}
+
+	// check just in case
+	if epoch.Validators == nil {
+		return false, fmt.Errorf("validators are not set for the current epoch")
+	}
+
+	msgVote := &MessageSignature{
+		From:      msg.NodeID,
+		Signature: msg.Signature,
+	}
+
+	if err := validateVote(msgVote, epoch); err != nil {
+		return false, err
+	}
+
+	numSignatures, err := c.state.insertMessageVote(msg.EpochNumber, msg.Hash, msgVote)
+	if err != nil {
+		return false, fmt.Errorf("error inserting message vote: %w", err)
+	}
+
+	c.logger.Info(
+		"deliver message",
+		"hash", hex.EncodeToString(msg.Hash),
+		"sender", msg.NodeID,
+		"signatures", numSignatures,
+		"quorum", getQuorumSize(len(epoch.Validators)),
+	)
+
+	return true, nil
+}
+
+func (c *consensusRuntime) runCheckpoint(epoch *epochMetadata) error {
+	// TODO: Implement checkpoint
+	return nil
+}
+
+// getLatestSprintBlockNumber returns latest sprint block number
+func (c *consensusRuntime) getLatestSprintBlockNumber() uint64 {
+	lastBuiltBlockNumber := c.lastBuiltBlock.Number
+
+	sprintSizeMod := lastBuiltBlockNumber % c.config.PolyBFTConfig.SprintSize
+	if sprintSizeMod == 0 {
+		return lastBuiltBlockNumber
+	}
+
+	sprintBlockNumber := lastBuiltBlockNumber - sprintSizeMod
+
+	return sprintBlockNumber
+}
+
+// calculateUptime calculates uptime for blocks starting from the last built block in current epoch,
+// and ending at the last block of previous epoch
+func (c *consensusRuntime) calculateUptime(currentBlock *types.Header) (*UptimeCounter, error) {
+	epoch := c.getEpoch()
+	uptimeCounter := &UptimeCounter{validatorIndices: make(map[ethgo.Address]int)}
+
+	if c.config.PolyBFTConfig.EpochSize < (uptimeLookbackSize + 1) {
+		// this means that epoch size must at least be 3 blocks,
+		// since we are not calculating uptime for lastBlockInEpoch and lastBlockInEpoch-1
+		// they will be included in the uptime calculation of next epoch
+		return nil, errors.New("epoch size not large enough to calculate uptime")
+	}
+
+	calculateUptimeForBlock := func(header *types.Header, validators AccountSet) error {
+		blockExtra, err := GetIbftExtra(header.ExtraData)
+		if err != nil {
+			return err
+		}
+
+		signers, err := validators.GetFilteredValidators(blockExtra.Parent.Bitmap)
+		if err != nil {
+			return err
+		}
+
+		for _, a := range signers.GetAddresses() {
+			uptimeCounter.AddUptime(ethgo.Address(a))
+		}
```

Comment:
Keep in mind one thing - you need _at least_ quorum validators to sign off on a block to commit to it. This means that you can include a block without all of the validators signing off on it - because you already received quorum for example.

This metric should probably not be used for making sure the validators are up

cc @JekaMas @b00ris @Kourin1996

Reply:
Completely aware of it (we don't have all the validators signatures captured this way, but only up to quorum size of signatures). 

We have outlined this to @ferranbt on one or two occasion, but this risk was accepted...